File: resilience-engine/internal/sim/simulation_test.go
package sim_test

import (
	"testing"

	"blueprint/resilience-engine/internal/model"
	"blueprint/resilience-engine/internal/sim"
)

func ecommerceSchema() model.SystemSchema {
	return model.SystemSchema{
		Nodes: []model.SystemNode{
			{EntityRef: "shop/web", Name: "Web App", Type: "web-app"},
			{EntityRef: "shop/mobile", Name: "Mobile App", Type: "mobile-app"},
			{EntityRef: "shop/api", Name: "API", Type: "microservice"},
			{EntityRef: "shop/payment", Name: "Payment", Type: "microservice"},
			{EntityRef: "shop/db", Name: "Database", Type: "database"},
		},
		Dependencies: []model.SystemDependency{
			{From: "shop/web", To: "shop/api"},
			{From: "shop/mobile", To: "shop/api"},
			{From: "shop/api", To: "shop/payment"},
			{From: "shop/api", To: "shop/db"},
		},
	}
}

func TestComputeBlastRadiusPropagatesUpstream(t *testing.T) {
	result := sim.ComputeBlastRadius(
		ecommerceSchema(),
		model.NodeFaultConfig{NodeID: "shop/payment", FaultType: "region-outage"},
		sim.BlastRadiusOptions{},
	)

	if result.Heat["shop/payment"] != 1 {
		t.Fatalf("expected payment heat 1, got %v", result.Heat["shop/payment"])
	}
	if result.Heat["shop/api"] <= 0 {
		t.Fatalf("expected api heat > 0")
	}
	if result.Heat["shop/web"] <= 0 {
		t.Fatalf("expected web heat > 0")
	}
	if _, ok := result.Heat["shop/db"]; ok {
		t.Fatalf("db should not be impacted")
	}
}

func TestRunSimulationDegradesSla(t *testing.T) {
	result := sim.RunSimulation(ecommerceSchema(), model.ChaosSpec{
		Faults: []model.NodeFaultConfig{
			{NodeID: "shop/payment", FaultType: "region-outage"},
		},
		EntryPoints: []string{"shop/web"},
	})

	if result.OverallSla >= 100 {
		t.Fatalf("expected degraded SLA, got %v", result.OverallSla)
	}
	if result.EntryPointSlas["shop/web"] >= 100 {
		t.Fatalf("expected degraded entry SLA")
	}
}

func TestDetectSpofs(t *testing.T) {
	spofs := sim.DetectSpofs(ecommerceSchema())
	found := false
	for _, id := range spofs {
		if id == "shop/api" {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected shop/api to be flagged as SPOF")
	}
}

func TestGroupDependencyExpansion(t *testing.T) {
	schema := model.SystemSchema{
		Nodes: []model.SystemNode{
			{EntityRef: "demo/user", Type: "person"},
			{EntityRef: "demo/hub", Type: "group", Name: "Hub"},
			{EntityRef: "demo/api", Type: "microservice", ParentEntityRef: "demo/hub"},
			{EntityRef: "demo/db", Type: "database", ParentEntityRef: "demo/hub"},
		},
		Dependencies: []model.SystemDependency{
			{From: "demo/user", To: "demo/hub"},
		},
	}

	result := sim.RunSimulation(schema, model.ChaosSpec{
		Faults: []model.NodeFaultConfig{
			{NodeID: "demo/api", FaultType: "region-outage"},
		},
		EntryPoints: []string{"demo/user"},
	})

	if result.Heat["demo/user"] <= 0 {
		t.Fatalf("expected user to be impacted via group edge, heat=%v", result.Heat["demo/user"])
	}
}

func TestRunMonteCarloProducesStats(t *testing.T) {
	stats := sim.RunMonteCarlo(ecommerceSchema(), model.ChaosSpec{
		Faults: []model.NodeFaultConfig{
			{NodeID: "shop/payment", FaultType: "latency", Severity: ptr(0.5)},
		},
		EntryPoints: []string{"shop/web"},
	}, model.MonteCarloConfig{Iterations: 200, Seed: 1})

	if stats.Iterations != 200 {
		t.Fatalf("expected 200 iterations, got %d", stats.Iterations)
	}
	if stats.OverallSlaMean <= 0 || stats.OverallSlaMean > 100 {
		t.Fatalf("unexpected mean SLA %v", stats.OverallSlaMean)
	}
	if stats.OverallSlaP95 < stats.OverallSlaP5 {
		t.Fatalf("p95 should be >= p5")
	}
}

func ptr(v float64) *float64 { return &v }

File: app/packages/core/src/resilience/simulation.test.ts
import { describe, expect, it } from 'vitest';
import type { SystemSchema } from '../models/schema';
import { detectSpofs } from './simulation';
import { runResilienceSimulation } from './simulation';

const ecommerceSchema: SystemSchema = {
  name: 'E-Commerce',
  version: '1.0.0',
  level: 'container',
  nodes: [
    { entityRef: 'shop/web', name: 'Web App', type: 'web-app' },
    { entityRef: 'shop/mobile', name: 'Mobile App', type: 'mobile-app' },
    { entityRef: 'shop/api', name: 'API', type: 'microservice' },
    { entityRef: 'shop/payment', name: 'Payment', type: 'microservice' },
    { entityRef: 'shop/db', name: 'Database', type: 'database' },
  ],
  dependencies: [
    { from: 'shop/web', to: 'shop/api', type: 'direct-call' },
    { from: 'shop/mobile', to: 'shop/api', type: 'direct-call' },
    { from: 'shop/api', to: 'shop/payment', type: 'direct-call' },
    { from: 'shop/api', to: 'shop/db', type: 'read-write' },
  ],
};

describe('runResilienceSimulation', () => {
  it('computes degraded SLA for entry points affected by blast radius', () => {
    const result = runResilienceSimulation(ecommerceSchema, {
      faults: [{ nodeId: 'shop/payment', faultType: 'region-outage' }],
      entryPoints: ['shop/web'],
    });

    expect(result.overallSla).toBeLessThan(100);
    expect(result.entryPointSlas['shop/web']).toBeLessThan(100);
    expect(result.heatHops.get('shop/payment')).toBe(0);
    expect(result.heatHops.get('shop/web')).toBeGreaterThan(0);
    expect(result.impactedDomains).toContain('shop');
    expect(result.advice.length).toBeGreaterThan(0);
  });

  it('reports full SLA when no faults are configured', () => {
    const result = runResilienceSimulation(ecommerceSchema, {
      faults: [],
      entryPoints: ['shop/web'],
    });

    expect(result.overallSla).toBe(100);
    expect(result.entryPointSlas['shop/web']).toBe(100);
  });

  it('merges blast radius across multiple simultaneous faults', () => {
    const single = runResilienceSimulation(ecommerceSchema, {
      faults: [{ nodeId: 'shop/payment', faultType: 'latency', severity: 0.5 }],
      entryPoints: ['shop/web'],
    });
    const dual = runResilienceSimulation(ecommerceSchema, {
      faults: [
        { nodeId: 'shop/payment', faultType: 'latency', severity: 0.5 },
        { nodeId: 'shop/db', faultType: 'error-rate', severity: 0.5 },
      ],
      entryPoints: ['shop/web'],
    });

    expect(dual.overallSla).toBeLessThanOrEqual(single.overallSla);
  });

  it('keeps entry-point SLA healthy when a publisher faults but marks integrity on async peers', () => {
    const schema: SystemSchema = {
      name: 'Pub-Sub',
      version: '1.0.0',
      level: 'container',
      nodes: [
        { entityRef: 'shop/web', name: 'Web', type: 'web-app' },
        { entityRef: 'shop/orders', name: 'Orders', type: 'microservice' },
        { entityRef: 'shop/events', name: 'Events', type: 'event-broker' },
        { entityRef: 'shop/worker', name: 'Worker', type: 'background-worker' },
      ],
      dependencies: [
        { from: 'shop/orders', to: 'shop/events', type: 'publish-subscribe' },
        { from: 'shop/worker', to: 'shop/events', type: 'publish-subscribe' },
      ],
    };

    const result = runResilienceSimulation(schema, {
      faults: [{ nodeId: 'shop/orders', faultType: 'region-outage' }],
      entryPoints: ['shop/web'],
    });

    expect(result.entryPointSlas['shop/web']).toBe(100);
    expect(result.overallSla).toBe(100);
    expect(result.integrityHeat.get('shop/worker')).toBeGreaterThan(0);
    expect(result.overallIntegrity).toBeLessThan(100);
    expect(result.advice.some(line => /miss new events/i.test(line))).toBe(true);
  });

  it('groups workspace-qualified refs by parent diagram, not workspace root', () => {
    const schema: SystemSchema = {
      name: 'Large Graph',
      version: '1.0.0',
      level: 'container',
      entityRef: 'blueprint/chaoslens-stress/large-graph',
      nodes: [
        {
          entityRef: 'blueprint/chaoslens-stress/large-graph/edge-mobile-01',
          name: 'Edge',
          type: 'web-app',
        },
        {
          entityRef: 'blueprint/chaoslens-stress/large-graph/bff-retail',
          name: 'BFF',
          type: 'rest-api',
        },
        {
          entityRef: 'blueprint/chaoslens-stress/large-graph/domain-orders',
          name: 'Orders',
          type: 'microservice',
        },
      ],
      dependencies: [
        {
          from: 'blueprint/chaoslens-stress/large-graph/edge-mobile-01',
          to: 'blueprint/chaoslens-stress/large-graph/bff-retail',
          type: 'direct-call',
        },
        {
          from: 'blueprint/chaoslens-stress/large-graph/bff-retail',
          to: 'blueprint/chaoslens-stress/large-graph/domain-orders',
          type: 'direct-call',
        },
      ],
    };

    const result = runResilienceSimulation(schema, {
      faults: [
        {
          nodeId: 'blueprint/chaoslens-stress/large-graph/domain-orders',
          faultType: 'region-outage',
        },
      ],
      entryPoints: ['blueprint/chaoslens-stress/large-graph/edge-mobile-01'],
    });

    expect(result.impactedDomains).toEqual(['large-graph']);
    expect(result.impactedDomains).not.toContain('blueprint');
  });
});

describe('detectSpofs', () => {
  it('flags shared dependencies with multiple callers and no circuit breaker', () => {
    const spofs = detectSpofs(ecommerceSchema);
    expect(spofs).toContain('shop/api');
    expect(spofs).not.toContain('shop/web');
  });

  it('excludes nodes that already have a circuit breaker safeguard configured', () => {
    const schema: SystemSchema = {
      ...ecommerceSchema,
      nodes: ecommerceSchema.nodes.map(n =>
        n.entityRef === 'shop/api'
          ? {
              ...n,
              resilience: { circuitBreaker: true },
            }
          : n
      ),
    };

    expect(detectSpofs(schema)).not.toContain('shop/api');
  });
});
